vllm/model_executor/models/aria.py

```diff
@@ -0,0 +1,1084 @@
+        config: AriaVisionConfig,
+        quant_config: Optional[QuantizationConfig] = None,
+        *,
+        prefix: str = "",
+    ) -> None:
+        super().__init__()
+
+        self.vision_model = AriaVisionTransformer(
+            config,
+            quant_config,
+            prefix=f"{prefix}.vision_model",
+        )
+
+    def forward(
+        self,
+        pixel_values: torch.Tensor,
+        pixel_mask: Optional[torch.BoolTensor] = None,
+    ) -> Tuple[torch.Tensor, Optional[torch.BoolTensor]]:
+        patch_attention_mask = self._create_patch_attention_mask(pixel_mask)
+
+        vit_oup = self.vision_model(
+            pixel_values=pixel_values,
+            patch_attention_mask=patch_attention_mask,
+        )
+
+        image_atts = self._create_image_attention_mask(patch_attention_mask)
+
+        return vit_oup, image_atts
+
+    def _create_patch_attention_mask(self, pixel_mask):
+        if pixel_mask is None:
+            return None
+
+        patches_subgrid = pixel_mask.unfold(
+            dimension=1,
+            size=self.vision_model.config.patch_size,
+            step=self.vision_model.config.patch_size,
+        ).unfold(
+            dimension=2,
+            size=self.vision_model.config.patch_size,
+            step=self.vision_model.config.patch_size,
+        )
+        return (patches_subgrid.sum(dim=(-1, -2)) > 0).bool()
+
+    def _create_image_attention_mask(self, patch_attention_mask):
+        if patch_attention_mask is None:
+            return None
+
+        flattened_mask = patch_attention_mask.flatten(1)
+        return torch.logical_not(flattened_mask)
+
+
+class FFN(nn.Module):
+    """
+    Feed-Forward Network module.
+
+    Args:
+        embed_dim (int): Input embedding dimension.
+        ff_dim (int): Hidden dimension of the feed-forward network.
+        output_dim (int): Output dimension.
+    """
+
+    def __init__(self, embed_dim, ff_dim, output_dim):
+        super().__init__()
+        self.linear_in = nn.Linear(embed_dim, ff_dim, bias=False)
+        self.linear_out = nn.Linear(ff_dim, output_dim, bias=False)
+        self.act = ACT2FN["gelu_new"]
+
+    def forward(self, hidden_states):
+        hidden_states = self.act(self.linear_in(hidden_states))
+        hidden_states = self.linear_out(hidden_states)
+        return hidden_states
+
+
+class CrossAttention(nn.Module):
+    """
+    Cross-Attention module.
+
+    Args:
+        kv_dim (int): Dimension of key and value.
+        embed_dim (int): Embedding dimension.
+        num_heads (int): Number of attention heads.
+        drop_out_rate (float): Dropout rate. Default is 0.
+    """
+
+    def __init__(self, kv_dim, embed_dim, num_heads, drop_out_rate=0):
+        super().__init__()
+        self.num_heads = num_heads
+        self.q_proj = nn.Linear(embed_dim, embed_dim, bias=False)
+        self.k_proj = nn.Linear(kv_dim, embed_dim, bias=False)
+        self.v_proj = nn.Linear(kv_dim, embed_dim, bias=False)
```

Comment:
> But I have a not very mature idea: perhaps we can use `ColumnParallelLinear` for q_proj and `MergedColumnParallelLinear` for kv_proj? Then we can use `RowParallelLinear` for `self.linear`.

I followed your suggestion. While this allowed inference to run without errors, the output is not correct.  Could you please take a look at [my changes](https://github.com/xffxff/vllm/commit/8d218f9b03d437f1cd4ce6e64fb08be28d60f547)?

Reply:
The issue might be about the `load_weights`:
```python3
    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]) -> Set[str]:
        stacked_params_mapping = [
            # (param_name, shard_name, shard_id)
            (".kv_proj", ".k_proj", 0),
            (".kv_proj", ".v_proj", 1),
        ]
        params_dict = dict(self.named_parameters())
        loaded_params: Set[str] = set()
        for name, loaded_weight in weights:
            for param_name, weight_name, shard_id in stacked_params_mapping:
                if weight_name not in name:
                    continue
                name = name.replace(weight_name, param_name)
                param = params_dict[name]
                weight_loader = param.weight_loader
                weight_loader(param, loaded_weight, shard_id)
                break
            loaded_params.add(name)
        return loaded_params
```
Seems that only the weight of `kv_proj` is loaded, perhaps you missed the "else" condition of `for ... else ...` statement?

Reply:
Ah, you're absolutely right! That was exactly the bug. Thanks a lot @Isotr0py 
(I didn't know about the `for ... else ...` syntax! That's surprising! Learned a lot today :)

Reply:
Hi @Isotr0py,

I just tested the code with TP enabled and encountered errors in this `CrossAttention` part. It seems that we also need to modify `nn.MultiHeadAttention`, which handles the q, k, and v projections internally. Adapting it to support TP is not that straightforward for me.

Since this attention is only used in a single layer, I’ve reverted the changes to the original implementation without considering TP support for now. We can optimize this later after this PR has landed. Is this okay?